src/api/routes/ws/web-socket-transmitter.ts

```diff
@@ -23,23 +24,45 @@ export class WebSocketTransmitter {
   readonly db: PgStore;
   readonly server: http.Server;
   private channels: WebSocketChannel[] = [];
+  private queue: PQueue;
 
   constructor(db: PgStore, server: http.Server) {
     this.db = db;
     this.server = server;
+    // This queue will send all messages through web socket channels, one at a time.
+    this.queue = new PQueue({
+      autoStart: true,
+      concurrency: 1,
+      timeout: 10_000, // 10 seconds
+      throwOnTimeout: true,
```

Comment:
I'm thinking through the implications of this, and it seems like a client (either malicious or just really slow network) could make a bunch of subscriptions that take a long time, say the full 10 second timeout, and effectively DDoS?

Reply:
Hmm that's a good point. The queue is a good way to protect itself against limited DDoS's in this way, but could still delay legitimate subscriptions from getting their updates in a major way. Perhaps the timeout should be shorter, like 1 second? Is there any update that would reasonably take longer than that?

Another change that would safeguard against this is to set a max number of subscriptions an API should logically have. However, without a shared state of socket subscriptions across all APIs, this would result in any new subscriptions over the max being denied. With a shared state of socket subscriptions across all APIs, they could redistribute the load and we could actually autoscale based on this metric. Food for thought in future designs perhaps.

Reply:
Good idea. I'm also going to double check the current server timeout behavior on WS and Socket IO i.e. there should be a timeout that automatically drops a client if a ping is not responded with a pong within a few seconds.

Reply:
It still sounds like I could craft an annoying client that responds to websocket pings in a timely fashion, but take a long time to receive each subscription message, then clog up all API deployments with a ton of subscriptions and result in no one getting subscription message. 

Maybe, in combination with your above changes, also limit the number of ws connections/subscriptions per client/IP?

Reply:
@zone117x check it out, this commit adds a per-emit timeout for socket.io... if 5 seconds are reached, the socket will be disconnected https://github.com/hirosystems/stacks-blockchain-api/pull/1427/commits/e76b2b1c608283ffa53cfba3ddb7cf09446a2e30